### PR TITLE
Session invalidation

### LINKED ISSUE #1652: Continue work started in PR #1593 to Implement server-side session invalidation using token_invalid_before
### Description

PR #1593 began work to do server side session invalidation. The original developer(s) will not be continuing this work. As a result, OED would like the work continued by others to address any comments and complete the task as needed.

If any comments, ideas or needed work is unclear or you have questions, then please reach out to the OED project.

### Additional context

New developers will not be able to push commits to the PR branch. The PR branch should be cloned to your machine, create a new branch and pushed to your own GitHub fork. Once the work is complete and on your GitHub fork, a new PR can be created from your branch on GitHub. The description should reference the original PR as the starting point and acknowledge the previous work's authors. Note OED requires that the original commit chain be present in the new PR and properly document who did what work. OED will close the original PR after any new one is merged. If you make progress but cannot address all needed work then OED would still like to have a draft PR created with the work you did complete. If you have questions about this process then please contact the OED project.

### PR DESCRIPTION
**Pull request by Better-Code-Saul-E**
# Description

Continue the work started in #1593 by @Oykunle  to implement server-side session invalidation for JWT authentication. This PR addresses the three outstanding review coments from #1593 that were left unresolved when the original author became unavailable to continue the work. See #1652.

Authors: @Oykunle  (original implementation), @Better-Code-Saul-E (review comment fixes)

1. Added a missing 'expect(beforeVerify.body).to.not.have.property('message')' assertion in the "token verify before logout" test
2. Added the missing 'message' property assertion on the first logout call in the "repeated logout" test
3. Renamed the 'login' variable in 'src/server/app.js' to 'loginLogout', and updated its usage in the same file to match the already renamed "loginLogout.js" route file

Fixes #1652 

## Type of change
- [x] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist
- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations
- As with the original implementation in #1593, this invalidates all tokens globally per user, not per-device or per-session.
- Requires a database migration for existing deployments

### CODE REVIEW

**Comment by huss** [comment]
@Better-Code-Saul-E Thank you for this contribution. The PR description states the CLA has been signed by all contributors. However, OED records do not have one for you. Thus, I would like you to sign the CLA via the link in the description or let me know that you think OED records are off.

Once this is done, this PR can be formally reviewed. Please let me know if you have any questions/thoughts.

**Comment by Better-Code-Saul-E** [comment]
I have signed the CLA.

File: src/server/app.js [hunk 1]
```diff
@@ -132,7 +132,7 @@ app.use('/api/users', users);
 app.use('/api/meters', meters);
 app.use('/api/readings', readings);
 app.use('/api/preferences', preferences);
-app.use('/api/login', login);
+app.use('/api/login', loginLogout);
```
Comment (huss) [inline src/server/app.js]: Your change above is good. The other part I wanted probably was not clear. Virtually every route in OED has the name the same but this one now has '/api/login' and 'loginLogout'. What I think would be better is to change it so it uses '/api/loginLogout' to match the new naming of the file so everything is consistent. This should mean that every place that uses the old name for the api needs to be updated. Once the change is done, it would be a very good idea to run all the tests (npm run test) and to try doing login/out operations via the OED web page to make sure everything works as desired.

**Review by huss — CHANGES_REQUESTED** [review]
@Better-Code-Saul-E Thank you for this contribution to OED. Your changes mostly addressed my previous comments. I've made one comment about a remaining item for you to consider. I have not carefully retested but will once all the changes are done. I see nothing you changed that should cause an issue but want to be careful. Please let me know if anything is not clear or you have thoughts/questions.

**Comment by Better-Code-Saul-E** [comment]
@huss I made the change. The manual login and automated tests worked!

Out of curiosity, besides consistency, is there a reason 'loginLogout' was preferred over something like 'auth' for the endpoint?

**Comment by huss** [comment]
> Out of curiosity, besides consistency, is there a reason 'loginLogout' was preferred over something like 'auth' for the endpoint?

Mostly it is historical since it was login so loginLogout seemed natural. auth isn't bad but I recall it may be used elsewhere as a term in OED to authenticate a user on the server when a request is made.

File: src/server/test/routes/authenticatorParamsTest.js [hunk 2]
```diff
@@ -213,7 +213,7 @@ mocha.describe('Authenticator Parameter Validation', () => {
 
 	mocha.describe('Security Edge Cases', () => {
 		mocha.it('should handle concurrent authentication attempts', async () => {
-			const LOGIN_ENDPOINT = '/api/login';
+			const LOGIN_ENDPOINT = '/api/loginLogout/login';
```
Comment (huss) [inline src/server/test/routes/authenticatorParamsTest.js]: You don't have to do this but every other usage uses the ``LOGIN_ENDPOINT`` that is set to this value. Same below. That would be nice and limit future changes.

File: .devcontainer/devcontainer-lock.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (huss) [inline .devcontainer/devcontainer-lock.json]: I think the entire file change should be removed from the PR. I suspect it automatically showed up on your system.

**Review by huss — CHANGES_REQUESTED** [review]
@Better-Code-Saul-E Thank you for addressing my previous comment. I have made two new ones to consider. Also, I forgot to mention last time that this PR is out-of-date with development (probably because it continues a previous, older PR). Now it has a merge conflict. Could you please fix this up so I can more fully test and merge the PR. Please let me know if you have any thoughts/questions.

**Comment by Better-Code-Saul-E** [comment]
@huss I resolved the merge conflict, removed devcontainer-lock.json, and fixed the duplicate LOGIN_ENDPOINT declaration in authenticatorParamsTest.js.

**Review by huss — COMMENTED** [review]
@Better-Code-Saul-E Thank you for the update. Review and testing found it is fine. However, the CI tests failed on GitHub. Unfortunately, the error message does not make the problem clear. I reran on my local machine and found it passed all tests. I reran them on GitHub and they failed again. I'm investigating as I have time and hope this isn't directly related to this PR. If anyone has thoughts/insights then they are more than welcome since I'm not an expert in this.

**Comment by Better-Code-Saul-E** [comment]
@huss I fixed the issue with the CI test failing. I added a missing license header in one of the sql files. The test passed. Could you verify on your machine when you get a chance?

**Comment by huss** [comment]
> @huss I fixed the issue with the CI test failing. I added a missing license header in one of the sql files. The test passed. Could you verify on your machine when you get a chance?

Thank you for figuring that out. I should have tried the check and not just the test command. I was doing too many things and it slipped my mind. They pass on my machine and GitHub so all is good.

**Review by huss — APPROVED** [review]
@Better-Code-Saul-E Review and testing found all is ready. Congratulations on your first accepted contribution to OED.
